File: packages/object-detection/ios/Plugin/ObjectDetectionPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ObjectDetectionPlugin)
public class ObjectDetectionPlugin: CAPPlugin {
    public let tag = "ObjectDetection"

    public let errorPathMissing = "path must be provided."
    public let errorLoadImageFailed = "The image could not be loaded."
    public let errorWriteFileFailed = "The result could not be created."

    public let defaultConfidence: Float = 0.9

    private var implementation: ObjectDetection?

    override public func load() {
        implementation = ObjectDetection(plugin: self)
    }

    @objc func processImage(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }

        let width = call.getInt("width")
        let height = call.getInt("height")

        let confidence = call.getFloat("confidence", defaultConfidence)

        guard let image = implementation?.createImageFromFilePath(path) else {
            call.reject(errorLoadImageFailed)
            return
        }

        let options = ProcessImageOptions(image: image,
                                          width: width,
                                          height: height,
                                          confidence: CGFloat(confidence))

        implementation?.processImage(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, nil, error)
                return
            }

            if let result = result {
                do {
                    call.resolve(try result.toJSObject())
                } catch {
                    call.reject(self.errorWriteFileFailed)
                }
            }
        })
    }
}

File: packages/object-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/objectdetection/ObjectDetectionPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.objectdetection;

import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.vision.common.InputImage;

import io.capawesome.capacitorjs.plugins.mlkit.objectdetection.classes.ProcessImageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.objectdetection.classes.ProcessImageResult;

@CapacitorPlugin(name = "ObjectDetector")
public class ObjectDetectionPlugin extends Plugin {

    public static final String TAG = "ObjectDetection";

    public static final String ERROR_PROCESS_IMAGE_CANCELED = "processImage canceled.";
    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_LOAD_IMAGE_FAILED = "The image could not be loaded.";

    public static final float CONFIDENCE = 0.9f;

    private MyObjectDetection implementation;

    @Override
    public void load() {
        try {
            implementation = new MyObjectDetection(this);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void processImage(PluginCall call) {
        try {
            String path = call.getString("path", null);
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }

            Integer width = call.getInt("width", null);
            Integer height = call.getInt("height", null);

            Float confidence = call.getFloat("confidence", CONFIDENCE);

            InputImage image = implementation.createInputImageFromFilePath(path);
            if (image == null) {
                call.reject(ERROR_LOAD_IMAGE_FAILED);
                return;
            }
            ProcessImageOptions options = new ProcessImageOptions(image, width, height, confidence);

            implementation.processImage(
                options,
                new ProcessImageResultCallback() {
                    @Override
                    public void success(ProcessImageResult result) {
                        try {
                            call.resolve(result.toJSObject());
                        } catch (Exception exception) {
                            String message = exception.getMessage();
                            Logger.error(TAG, message, exception);
                            call.reject(message);
                        }
                    }

                    @Override
                    public void cancel() {
                        call.reject(ERROR_PROCESS_IMAGE_CANCELED);
                    }

                    @Override
                    public void error(Exception exception) {
                        String message = exception.getMessage();
                        Logger.error(TAG, message, exception);
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            String message = exception.getMessage();
            Logger.error(TAG, message, exception);
            call.reject(message);
        }
    }
}